strongbox-db-server/src/main/java/org/strongbox/db/server/janusgraph/TransactionalVertexIDAssigner.java

```diff
@@ -0,0 +1,157 @@
+import org.janusgraph.graphdb.database.idassigner.VertexIDAssigner;
+import org.janusgraph.graphdb.idmanagement.IDManager;
+import org.janusgraph.graphdb.internal.InternalRelation;
+import org.janusgraph.graphdb.internal.InternalVertex;
+
+/**
+ * @author sbespalov
+ */
+public class TransactionalVertexIDAssigner extends VertexIDAssigner
+{
+
+    public static final String QUEUE_DEFAUT = "default";
+    public static final String TX_RESOURCE_QUEUE = TransactionalVertexIDAssigner.class.getName() + ".queue";
+
+    private Map<String, VertexIDAssigner> queueMap = new ConcurrentHashMap<>();
+    private final Function<String, VertexIDAssigner> newInstanceFactory;
+    private final Supplier<String> idBlockQueueSupplier;
+
+    public TransactionalVertexIDAssigner(Configuration config,
+                                         IDAuthority idAuthority,
+                                         StoreFeatures idAuthFeatures,
+                                         Supplier<String> idBlockQueueSupplier)
+    {
+        super(config, idAuthority, idAuthFeatures);
+        newInstanceFactory = (queueId) -> new VertexIDAssigner(config, new ReadOnlyIDAuthority(idAuthority), idAuthFeatures);
+        this.idBlockQueueSupplier = idBlockQueueSupplier;
+    }
+
+    @Override
+    public IDManager getIDManager()
+    {
+        return super.getIDManager();
+    }
+
+    @Override
+    public synchronized void close()
+    {
+        queueMap.values().forEach(VertexIDAssigner::close);
+    }
+
+    @Override
+    public void assignID(InternalRelation relation)
+    {
+        VertexIDAssigner currentIDAssigner = getCurrentIDAssigner();
+        if (currentIDAssigner == this)
+        {
+            super.assignID(relation);
+        }
+        else
+        {
+            currentIDAssigner.assignID(relation);
+        }
+    }
+
+    @Override
+    public void assignID(InternalVertex vertex,
+                         VertexLabel label)
+    {
+        VertexIDAssigner currentIDAssigner = getCurrentIDAssigner();
+        if (currentIDAssigner == this)
+        {
+            super.assignID(vertex, label);
+        }
+        else
+        {
+            currentIDAssigner.assignID(vertex, label);
+        }
+    }
+
+    @Override
+    public void assignIDs(Iterable<InternalRelation> addedRelations)
+    {
+        VertexIDAssigner currentIDAssigner = getCurrentIDAssigner();
+        if (currentIDAssigner == this)
+        {
+            super.assignIDs(addedRelations);
+        }
+        else
+        {
+            currentIDAssigner.assignIDs(addedRelations);
+        }
+    }
+
+    protected VertexIDAssigner getCurrentIDAssigner()
+    {
+
+        return getCurrentQueueId().map((id) -> queueMap.computeIfAbsent(id, newInstanceFactory)).orElse(this);
+    }
+
+    private Optional<String> getCurrentQueueId()
+    {
+        return Optional.ofNullable(idBlockQueueSupplier.get());
+    }
+
+    private class ReadOnlyIDAuthority implements IDAuthority {
```

Comment:
Could you please add a comment above this class explaining why it's necessary and why it's `ReadOnly`? Thanks! :)